htmresearch/algorithms/hierarchical_clustering.py

```diff
@@ -0,0 +1,113 @@
+from nupic.algorithms.KNNClassifier import KNNClassifier
+
+
+
+class HierarchicalClustering(object):
+  """
+  Implements hierarchical agglomerative clustering on the output of a classification network.
+  """
+
+
+  def __init__(self, classificationNetwork, linkageMethod='single'):
+    knn = classificationNetwork.getKNN()
+    dim = len(knn.getPattern(0, sparseBinaryForm=False))
+    sparseRowList = []
+
+    for i in xrange(knn._numPatterns):
+      nzIndices = knn.getPattern(i, sparseBinaryForm=True)
+      sparseRow = scipy.sparse.csr_matrix(
+        (numpy.ones(len(nzIndices), dtype=bool),
+        (numpy.zeros(len(nzIndices)), nzIndices)),
+        shape=(1,dim))
+      sparseRowList.append(sparseRow)
+
+    sparseDataMatrix = scipy.sparse.vstack(sparseRowList)
+    overlaps = HierarchicalClustering._computeOverlaps(sparseDataMatrix)
+
+    print overlaps.nbytes / 1024**2 # MB
+
+    linkage = scipy.cluster.hierarchy.linkage(-overlaps, method=linkageMethod)
+    
+    linkage[:,2] -= numpy.min(linkage[:,2])
+    print numpy.max(linkage[:,2])
+    self._linkage = linkage
+
+
+  @staticmethod
+  def _computeOverlaps(data, selfOverlaps=False, dtype='int16'):
+    """
+    Calculates all pairwise overlaps between the rows of the input. Returns an
+    array of with all n(n-1)/2 values in the upper triangular portion of the
+    pairwise overlap matrix. Values are return in row-major order.
+
+    @param data (scipy.sparse.csr_matrix) A CSR sparse matrix with one vector
+        per row. Any non-zero value is considered an active bit.
+
+    @param selfOverlaps (boolean) If true, include diagonal (density) values
+        from pairwise similarity matrix. Then returned vector has n(n+1)/2
+        elements. Optional, defaults to False.
+    
+    @param dtype (string) Data type of returned array in numpy dtype format.
+        Optional, defaults to 'int16'.
+    
+    @returns (numpy.ndarray) A vector of pairwise overlaps as described above
+        with dtype uint16.
+    """
+    nVectors = data.shape[0]
+    nDims = data.shape[1]
+    nPairs = (nVectors+1)*nVectors/2 if selfOverlaps else (
+        nVectors*(nVectors-1)/2)
+    overlaps = numpy.ndarray(nPairs, dtype=dtype)
+    pos = 0
+
+    for i in xrange(nVectors):
+      start = i if selfOverlaps else i+1
+      a = data[i]
+      b = data[start:]
+      newOverlaps = a.multiply(b).getnnz(1)
+      run = newOverlaps.shape[0]
+      overlaps[pos:pos+run] = newOverlaps
+      pos += run
+    return overlaps
+
+
+  def getLinkageMatrix(self):
+    """
+    Returns a linkage matrix of the form defined by
+    http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.cluster.hierarchy.linkage.html
+    """
+    return self._linkage
+
+
+  def getDendrogram(self):
+    linkage = self.getLinkageMatrix()
+    fig = plt.figure()
+    ax = plt.axes()
+    scipy.cluster.hierarchy.dendrogram(linkage, p=30, truncate_mode="lastp", ax=ax)
+    return fig
+    
+
+
+if __name__ == "__main__":
```

Comment:
I would recommend moving this code block to projects/nlp/clustering as simple_clustering_example.py